core/src/main/java/io/questdb/griffin/engine/functions/groupby/CountSymbolGroupByFunction.java

```diff
@@ -32,15 +32,15 @@
 import io.questdb.griffin.engine.functions.GroupByFunction;
 import io.questdb.griffin.engine.functions.LongFunction;
 import io.questdb.griffin.engine.functions.UnaryFunction;
-import io.questdb.std.IntList;
+import io.questdb.std.CompactIntHashSet;
 import io.questdb.std.Numbers;
 import io.questdb.std.ObjList;
 
 import static io.questdb.cairo.sql.SymbolTable.VALUE_IS_NULL;
 
 public class CountSymbolGroupByFunction extends LongFunction implements UnaryFunction, GroupByFunction {
     private final Function arg;
-    private final ObjList<IntList> lists = new ObjList<>();
+    private final ObjList<CompactIntHashSet> sets = new ObjList<>();
```

Comment:
It may be enough to have 1 `CompactLongHashSet` and store there 2 integers as 1 long : group_id, symbol_id

Reply:
`ObjList<CompactIntHashSet>` wins by factor of 2 when there is > 10 items in the group, otherwise it loses because of min size. It also can create 1000s of heap objects.

Reply:
Now I get what you're proposing. I'd better create a `FastSet` similar to what we have in `FastMap` and use it in all `count_distinct()` implementations for fixed-size types. WDYT?

Reply:
Replaced all `Compact*HashSet` sets and `Long256HashSet` with `FastMap` in 35870a06b459957fae88757be32c518f94bf8abf

Going to measure performance now. A side note: #2868 will improve performance of `FastMap`, so that's something to keep in mind.

Reply:
Here are the results:
```sql
create table x as (
  select rnd_symbol(100, 10, 10, 0) sym1,
         rnd_symbol(100, 10, 10, 0) sym2,
         rnd_symbol(181000, 10, 10, 0) sym3,
         cast(172800000000 as timestamp) ts
  from long_sequence(1000000)
) timestamp(ts) partition by day;

select count_distinct(sym3) from x; -- 100ms vs. 30ms before FastMap

select ts, sym1, sym2, count_distinct(sym3) from x; -- 250ms vs 180ms before FastMap
```

So, 3-4x worse times. I've checked if there is any clustering (hash code collisions), but in this scenario there are none. It seems that we spend most of the time comparing keys. See the below flame graph collected while running the first query.

So far, I don't think `eq()` is very inefficient and there are a lot of opportunities to improve it significantly. Looks like we need to decide whether we want to pay the price of on-heap allocations (GC pressure and potential OOMEs) and better performance when the data volume is small or we want to use `FastMap` in `count_distinct()` and, maybe, try to optimize it.

cc @jerrinot who was also complaining about `FastMap` recently.

![Screenshot from 2023-01-06 11-44-19](https://user-images.githubusercontent.com/37772591/210974972-acd2a067-7030-403a-9506-c755fc87c63b.png)

Reply:
Sounds like `ObjList<CompactIntHashSet>` is better than `FastMap`. Why `FastMap`, not `CompactLongHashSet`?

From where it stands now, first implementation iteration was better than the second.

What's your opinion, you think `FastMap` is good enough or you want to try something else? Binary Search Tree can be a candidate, it's compact and insertion / search is logarithmic but we will need to write the data structure, which will take quite some time.

Reply:
> Why `FastMap`, not `CompactLongHashSet`?

That's because `FastMap` applies to all `count_distinct()` overloads, not just int and symbol. The main advantage is that the data is stored off-heap instead of lots of on-heap objects.

> What's your opinion, you think `FastMap` is good enough or you want to try something else? Binary Search Tree can be a candidate

In this particular test there were no collisions, so a binary tree would have no advantage. The time seems to be spend mostly on key equality check which is inevitable with search in any data structure. We might want to spend some time optimizing `FastMap` once all in-flight PRs are merged.

I'm going to revert `FastMap`-related commits for now as this fix is important for one of our users.

Reply:
Reverted the changes in 5536c9a96b2d78d978e45477f1aa37146aaae4e7 and 1b45cb7941a851bf9be5ebae670611967517ef57

Going to give a try optimizing `FastMap` in a separate PR, probably next week.